Reimplement recipe entry as a modal dialog

gnarlyquack opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Describe your changes in detail -->
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
This re-implements the recipe editor as a modal dialog while also simplifying much of the UI logic and eliminating the use of deprecated GTK components. I think this fixes #244, #57, and #256. This does break plugin support, however the existing infrastructure relies heavily on UIManagers, which [are deprecated](https://lazka.github.io/pgi-docs/Gtk-3.0/classes/UIManager.html#Gtk.UIManager). That said, I believe the only two plugins impacted are the UnitConverter, which can probably be more effectively integrated into the dialog, and the KeyEditor, which can probably be eliminated with some work on the database.

This definitely needs some testing and pounding on by others. Looking forward to feedback.

## How Has This Been Tested?
<!--- Please describe in detail how you tested your changes. -->
<!--- Include details of your testing environment, tests ran to see how -->
<!--- your change affects other areas of the code, etc. -->
By exercising the UI: adding, editing, deleting, and cancelling edits to recipes.

## Screenshots (if appropriate):

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

cydanil commented:
It seems that the issue within the CI is that the `gtk`folder you've added should also be added to the packages in `setup.py`. Why not use the `ui` folder?

gnarlyquack commented:
I haven't taken a look at the build issues yet. I can probably get to them later today.

I ditched the `.ui` file because I needed to rework the ingredients page and found it simpler to just do in code. If I may say so, I think it also ends up being more straightforward.

cydanil commented:
I also think that pure python is more maintainable. I wonder why not put the python code within the `ui` folder and import from `gourmet.ui.recipe` instead of `gourmet.gtk.recipe`

gnarlyquack commented:
Oh, I misunderstood your question. No real reason, other than the `ui` folder currently only contains `.ui` files and I wanted to avoid mixing source code and resources. However, I did want to start separating GTK-specific code from "domain" logic and isolate it into its own package, and calling it `gtk` seemed like a natural choice. In theory, somebody could then implement another front end by porting the `gtk` package to another toolkit.

cydanil reviewed (changes requested):
This PR brings a lot of welcome changes, and I thank you for your work!

Please
- add `po/.intltool-merge-cache.lock` to the gitignore and remove it from the PR
- move `test_recipe.py` to `gourmet/tests/dogtail`.
- make the test run with dogtail (This could be done in a different PR, if you feel that it would take too long now).

In another PR, the machinery to run dogtail tests will be added.

gnarlyquack commented:
`test_recipe.py` doesn't actually have any tests in it, it's just a script to launch the dialog in isolation so it can be manually tested. I wasn't sure what else to call it. This is similar to the `if __name__ == '__main__'` pattern used in much of the code to launch UI components in isolation for manual testing, but running `recipe.py` directly was causing some breakages (specifically with `isinstance`) since the module was being identified as `__main__` instead of `gourmet.gtk.recipe`. I feel like we probably want to be able to launch UI components in isolation, and we probably also want to move this code out of the modules themselves so we don't distribute it in wheels, etc. Perhaps we want another directory (separate from tests) to stick these scripts?

I can write some dogtail tests, but if you're otherwise ready to merge this PR, I'd prefer to add them as a separate PR if it's the same to you.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+import os as _os
+
+from gi.repository import (
+    Gdk as _Gdk,
```
What's with the `_`-prefixed imports and function definitions?

gnarlyquack replied:
[Per PEP 8](https://www.python.org/dev/peps/pep-0008/#id36), this signifies internal use and prevents the symbol from being imported when doing `from x import *`.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+
+    # TODO: Eliminate dependency on RecGui
+    def run(self, parent, rec_gui, recipe=None, module=None):
+        # TODO: preferences?
```
It is somewhat unclear what this comment refers to

gnarlyquack replied:
The current recipe entry UI has a menu with a preferences option. This is a reminder to look at what the preferences were managed by that option and see if any functionality needs to be replicated.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+            # This uses the same logic as the original code, but should this
+            # ever fail?
+            try:
+                yields = float(recipe['yields'])
```
Yes, the try/except can be removed.

gnarlyquack replied:
Are we sure? `yields` might be `None`, which could raise an error. The TODO is meant as a reminder/placeholder to take a closer look at this.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    button_remove.show()
+
+
+def _generate_thumbnail(widget):
```
These three helper functions below should be unit-tested, to ensure that there's no breakage.  

They are also duplicating some functionalities from `image_utils.py`. Perhaps they could be merged?

gnarlyquack replied:
I thought I had eliminated duplication with `image_utils`, although `image_utils` may perform the same functionality but with PIL. It appears that GDK Pixbufs may offer equivalent functionality, and since we already depend on the GObject infrastructure, using GDK directly means we could potentially eliminate intermediate conversions and drop the dependency on PIL altogether.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    # dragged from instead of staying on the dropped row, which is visually
+    # quite confusing: it makes it difficult to determine where the row was
+    # dropped and even whether or not the operation succeeded. So we have to
+    # take care of this ourselves. Of course, GTK doesn't make it easy.
```
```suggestion
    # take care of this ourselves.
```

I also have issues with Gtk. However, the code should remain neutral

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    # dropped and even whether or not the operation succeeded. So we have to
+    # take care of this ourselves. Of course, GTK doesn't make it easy.
+    # Drag-and-drop is essentially implemented by deleting the row at the drag
+    # point and inserting a new row at the drop point. You might think you
+    # could just connect to the 'row-inserted' signal to select the new row
+    # when it's inserted, but GTK does everything in the opposite order you
+    # might expect: it first inserts a blank row at the drop point -- which is
```
```suggestion
    # point and inserting a new row at the drop point. GTK first inserts a 
    # blank row at the drop point -- which is
```
I think that comments should remain neutral and factual.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    column = _Gtk.TreeViewColumn(_('Optional'), renderer, active=3)
+    tree.append_column(column)
+
+    # Prevent collapsing ingredient groups. This seems preferable to setting
```
That, I like :)

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    dialog.show_all()
+    response = dialog.run()
+
+    if _Gtk.ResponseType.OK == response:
```
That's a trick that I do a lot in C/C++, but PEP8 enforces 
```suggestion
    if response ==_Gtk.ResponseType.OK:
```
It also applies to other occurrences throughout this file.

gnarlyquack replied:
I'm not getting any errors about this from pycodestyle (formerly pep8). The rest is whitespace and line length issues which I can certainly fix.

cydanil replied:
I use `flake8`. I stand corrected though: it seem to be valid!   
I would still ask you to stick to the more conventional `val == const`

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+    ingredients.ingredient_entry.grab_focus()
+
+
+def _check_if_recipe(db, name):
```
It's debatable whether a recipe title can be used as a unique ID. 
It's certainly interesting, but could break user's workflow.  
Have you tried to have a database that already contains 2 recipes with the same title? I can think of a number of scenarios where it might be the case (eg. imports from backups, from other applications).  

I wonder what would happen if a user had their database with duplicates and would update to a Gourmet with this.

gnarlyquack replied:
There's no enforcement of this from either the UI or the database, but recipe names _should_ be unique, even if the user should want to enter multiple recipes for the same dish (e.g., multiple recipes for "brownies"), otherwise they have no way to distinguish different recipes (other than reading the recipe). Or, alternatively, enforcing some unique combination of recipe name and, say, source. This does require adding a unique key to the database and adding an upgrade step that ensures the key is satisfied. While this might be a short-term or in-the-moment inconvenience for users, in the long run it should reduce duplication, produce a better-quality recipe database, and enable more powerful features. Ideally, this change should happen in conjunction with a refactoring of the database.

zombiewoof replied:
I have noticed, in export data, a line for each recipe (example):

	<recipe id="1247">

Is that not a unique identifier stored in the database?

If it doesn't exist already, I think that adding a unique key to the database is a good idea, even if it requires an upgrade step for users. I also agree that names should be unique, but don't think that is a good unique key. A problem may arise if users' current database has duplicated names. It would be a really good idea to give the user a warning if they are attempting to add/edit a recipe where the name would be duplicated - and disallow save until the name is unique.

gnarlyquack replied:
There is a unique integer key in the database which is used as the primary key, which we will certainly want to keep as [a surrogate key](https://en.wikipedia.org/wiki/Surrogate_key), however there's no enforcement of [the natural key](https://en.wikipedia.org/wiki/Natural_key), which is how users distinguish different recipes. I would suggest using the name of the recipe as the natural key and putting a uniqueness constraint on it. Right now, there's nothing preventing a user from entering different recipes with the same name, and in fact Gourmet currently bundles a plugin to find duplicate recipes. Now certainly we want users to be able to enter multiple recipes for something, but we should make sure the user is able to somehow distinguish the different recipes (and probably alert them that they already have a similar recipe), and the easiest way would be to require a unique recipe name. As an example from the world of cocktails, notice how many different Cosmopolitan cocktail recipes are listed, but they're all distinguished by name: https://www.diffordsguide.com/search?q=cosmopolitan

The fact that duplicate recipe names have been allowed until now means we will need some sort of database upgrade step to sort this out. It can either specifically prompt the user for unique names, which may not be practical if there are a large number of duplicates in the database, or we can algorithmically make the names unique (which could be as simple as adding "1", "2", etc. to the end of the name).

zombiewoof replied:
Ah, I understand what you're saying, now. To be honest, I've never understood why duplicate names were allowed. Duplicated names, as you say, make it more difficult to differentiate between similar recipes.

I lean toward doing it algorithmically to make the upgrade transition as quick as possible.

Maybe provide the user with a list of renamed recipes (old and new names) - either a text file or something they can copy/paste into a document to save for later use. If not providing the data, as described, make the rename add some text to the new name that the user can search for. For example, two recipes named Brownies could be renamed as 'Brownies renamed 1' and 'Brownies renamed 2'. Either method would allow the user to go in, at their leisure, and fix any names that were previously duplicated.

cydanil commented on gourmet/gtk/recipe.py:
```diff
@@ -0,0 +1,787 @@
+
+def _select_image(image, widget, button_add, button_remove):
+    filename = _select_image_dialog("Select Image",
+                                action=_Gtk.FileChooserAction.OPEN)
```
indent?

cydanil commented on test_recipe.py:
```diff
@@ -0,0 +1,44 @@
+import sys
```
This file should be moved elsewhere than at the root of the project, ideally to `tests/dogtail` or `tests/broken`

gnarlyquack replied:
As I mentioned in my earlier comment, this file really isn't a "test" and simply a script to launch the UI on its own so it can be manually tested.

cydanil commented on gourmet/reccard.py:
```diff
@@ -1,47 +1,33 @@
+from gourmet.plugin import RecDisplayPlugin, ToolPlugin
+
+from gourmet.gtk.recipe import (
+    RecipeEditor as _RecipeEditor,
```
I'm unsure about the `_` prefix.

gnarlyquack replied:
See above.

cydanil reviewed (changes requested):
Code-wise, it looks cool, and it feels snappier.  
However, many features have been removed, rich text and undo/redo most notably:

![image](https://user-images.githubusercontent.com/2159577/97235103-751fd400-17e2-11eb-9970-b982a0551e1c.png)


I suggest to create a flatpak, and ask one of @zombiewoof, @Oldbwl, or @allenerenee for feedback.

I also get 86 PEP8 violations and a bunch of AttributeErrors when editing a recipe...

gnarlyquack commented:
> Code-wise, it looks cool, and it feels snappier.
> However, many features have been removed, rich text and undo/redo most notably:

Since the recipe editor is now implemented using a standard modal dialog paradigm, changes are no longer committed until the entire dialog is accepted, and any mistakes can be undone by cancelling the dialog. I also removed the various menus which aren't typical of dialogs (they were also buggy; see #244, for example). For this reason, I removed undo/redo.

I also removed the PangoBuffer controls as GTK in general and the Python bindings in particular don't really give us any tools to work with Pango markup other than parsing the text buffer when reading it out, which seems hackish to me. Other projects have dropped this as well as they've moved to GTK 3, see: https://github.com/FluidityProject/spud/pull/16 and https://github.com/pitivi/pitivi/commit/27cfba0e1e4b7d16aafef6fa1ad0fef9e35584c1 as some examples. That said, a toolbar can be added relatively easily to the Instructions and Notes pages to re-enable text formatting.

> I also get 86 PEP8 violations and a bunch of AttributeErrors when editing a recipe...

After installing pycodestyle (formerly pep8), the only PEP 8 violations I'm seeing are whitespace and line length, which I can certainly fix.

The attribute errors are from the UI Plugin system, which is expecting a ui_manager attribute. I've removed the UIManager since GTK has deprecated the component. If I the recipe editor doesn't inherit from `Pluggable` then the errors will disappear.

cydanil commented:
I think that you ought to rebase and then you or I create a Flatpak, such that we all can use the updated editor and validate the usability. From then on, we can further discuss the uniqueness of recipes.

cydanil commented:
I'm sorry that we could not reach an agreement :(  
Despite our differences, I think that you did a pretty good job of clearing up and modernising the code, and it certainly brought value in.
